flocker/acceptance/test_restarts.py

```diff
@@ -0,0 +1,169 @@
+
+        def query_server():
+            req = get(
+                "http://{host}:12345".format(host=node.public_address),
+                persistent=False
+            ).addCallbacks(content)
+            return req
+
+        def created_dataset(dataset):
+            print "DATASET", dataset
+            http_server = {
+                u"name": random_name(self),
+                u"node_uuid": str(node.uuid),
+                u"image": u"python:2.7-slim",
+                u"ports": [{u"internal": 12345, u"external": 12345}],
+                # We expect containers to be restarted irrespective of restart
+                # policy because FLOC-3137 makes that the intended behaviour.
+                # The restart policy will be ignored by the containers API.
+                u'restart_policy': {u'name': u'never'},
+                u"volumes": [{u"dataset_id": str(dataset.dataset_id),
+                              u"mountpoint": u"/data"}],
+                u"command_line": [u"python", u"-c",
+                                  REBOOT_SERVER.getContent().decode("ascii"),
+                                  u"/data"],
+            }
+            created = cluster.create_container(http_server)
+            created.addCallback(lambda _: self.addCleanup(
+                cluster.remove_container, http_server[u"name"]))
+            created.addCallback(lambda _:
+                                loop_until(lambda: query_server().addErrback(
+                                    lambda _: False)))
+            created.addCallback(lambda _: query_server())
+            return created
+        creating_dataset.addCallback(created_dataset)
+
+        def server_started(initial_response):
+            # We now have the initial response of the server. This should
+            # be the initial boot time of the node, repeated twice.  Next,
+            # we restart the machine.
+            print "INITIAL RESPONSE", initial_response
+            initial_reboot_time = initial_response.splitlines()[0]
+            initial_container_id = initial_response.splitlines()[2].encode("ascii")
+            # XXX Checking exit code is problematic insofar as reboot
+
+            # kills the ssh process...
+            print "Rebooting!"
+            disabling = _service(node.public_address, b'flocker-dataset-agent', b'disable')
+            def reboot(ignored):
+                call([b"ssh", b"root@{}".format(node.public_address),
+                      b"shutdown", b"-r", b"now"])
+            rebooting = disabling.addCallback(reboot)
+
+            # Now, keep trying to get a response until it's different than
+            # the one we originally got, thus indicating a reboot:
+            def query_until_different():
+                print "query!"
+                return query_server().addCallbacks(
+                    lambda response: response != initial_response,
+                    lambda _: False)
```

Comment:
Oh wait, I'm confused, this doesn't get called yet. It might be clearer if we defined this function immediately before it gets assigned to the deferred.